scripts/merge-script/pull-request.ts

```diff
@@ -0,0 +1,128 @@
+/**
+ * @license
+ * Copyright Google LLC All Rights Reserved.
+ *
+ * Use of this source code is governed by an MIT-style license that can be
+ * found in the LICENSE file at https://angular.io/license
+ */
+
+import * as Octokit from '@octokit/rest';
+import {GitClient} from './git';
+import {PullRequestMergeTask} from './index';
+import {matchesPattern} from './string-pattern';
+import {getBranchesFromTargetLabel, getTargetLabelFromPullRequest} from './target-label';
+
+/** Enum that describes different states of a pull request. */
+export const enum PullRequestStatus {
+  CLA_UNSIGNED,
+  NOT_MERGE_READY,
+  FAILING_CI_JOBS,
+  PENDING_CI_JOBS,
+  NO_TARGET_LABEL,
+  MERGE_CONFLICTS,
+  UNKNOWN_MERGE_ERROR,
+  UNSATISFIED_BASE_SHA,
+  NOT_FOUND,
+  MISMATCHING_TARGET_BRANCH,
+}
+
+/** Interface that describes a pull request. */
+export interface PullRequest {
+  /** Number of the pull request. */
+  prNumber: number;
+  /** Title of the pull request. */
+  title: string;
+  /** Labels applied to the pull request. */
+  labels: string[];
+  /** List of branches this PR should be merged into. */
+  targetBranches: string[];
+  /** Branch that the PR targets in the Github UI. */
+  githubTargetBranch: string;
+  /** Amount of commits this pull request contains. */
+  commitsAmount: number;
+  /** Optional SHA that this pull request needs to be based on. */
+  requiredBaseSha?: string;
+  /** Whether the pull request commit message fixup. */
+  needsCommitMessageFixup: boolean;
+}
+
+/** Failure for pull request. */
+export interface PullRequestFailure {
+  /** Status of the associated pull request. */
+  status: PullRequestStatus;
+  /**
+   * Optional context for the pull request failure. e.g. Name of the branch that
+   * the pull request could not be cherry-picked into.
+   */
+  context?: any;
```

Comment:
I think to do so, we would need to overload this interface for each of the potential `PullRequestStatus`es, I am not entirely should how much value it would add.  But could definitely be considered.

Reply:
Yeah, I agree that this is not really ideal. My goal here is that the actual merge logic does not have to deal with writing messages (also since that avoids duplication).

I'm changing it to an abstract class with static methods on it. e.g.

```ts
return PullRequestFailure.mergeConflicts(failedBranches)
```

Let me know if that works for you all.